examples/energy-management-app/energy-management-common/src/DeviceEnergyManagementDelegateImpl.cpp

```diff
@@ -0,0 +1,397 @@
+    {
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+
+    mEsaState = newValue;
+    if (oldValue != newValue)
+    {
+        ChipLogDetail(AppServer, "mEsaState updated to %d", static_cast<int>(mEsaState));
+        MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, ESAState::Id);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceEnergyManagementDelegate::SetAbsMinPower(int64_t newValue)
+{
+    int64_t oldValue = mAbsMinPower;
+
+    mAbsMinPower = newValue;
+    if (oldValue != newValue)
+    {
+        ChipLogDetail(AppServer, "mAbsMinPower updated to %d", static_cast<int>(mAbsMinPower));
+        MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, AbsMinPower::Id);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceEnergyManagementDelegate::SetAbsMaxPower(int64_t newValue)
+{
+    int64_t oldValue = mAbsMaxPower;
+
+    mAbsMaxPower = newValue;
+    if (oldValue != newValue)
+    {
+        ChipLogDetail(AppServer, "mAbsMaxPower updated to %d", static_cast<int>(mAbsMaxPower));
+        MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, AbsMaxPower::Id);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR
+DeviceEnergyManagementDelegate::SetPowerAdjustmentCapability(PowerAdjustmentCapability::TypeInfo::Type powerAdjustmentCapability)
+{
+    if (powerAdjustmentCapability.IsNull())
+    {
+        mPowerAdjustmentCapability.SetNull();
+        ChipLogDetail(AppServer, "DEM: PowerAdjustmentCapability is null");
+    }
+    else
+    {
+        auto src    = powerAdjustmentCapability.Value();
+        auto target = mPowerAdjustmentCapability.Value();
+        auto size   = src.size();
+
+        if (size > 8)
+        {
+            return CHIP_ERROR_BAD_REQUEST;
+        }
+
+        target.reduce_size(0);
+
+        for (size_t i = 0; i < size; i++)
+        {
+            const_cast<Structs::PowerAdjustStruct::Type &>(target[i]) = src[i];
+        }
+    }
+
+    MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, PowerAdjustmentCapability::Id);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceEnergyManagementDelegate::SetForecast(DataModel::Nullable<Structs::ForecastStruct::Type> forecast)
+{
+    if (forecast.IsNull())
+    {
+        mForecast.SetNull();
+        ChipLogDetail(AppServer, "DEM: Forecast is null");
+    }
+    else
+    {
+        auto src    = forecast.Value();
+        auto target = mForecast.Value();
+
+        target.forecastId++;
+        target.activeSlotNumber  = src.activeSlotNumber;
+        target.startTime         = src.startTime;
+        target.endTime           = src.endTime;
+        target.earliestStartTime = src.earliestStartTime;
+        target.latestEndTime     = src.latestEndTime;
+        target.isPauseable       = src.isPauseable;
+
+        auto slots = src.slots;
+        auto size  = slots.size();
+        target.slots.reduce_size(0); // clear them
```

Comment:
same here:  you seem to want reduce_size (size)... but then you need to ensure that existing size is large enough.

Reply:
I've removed this code since it needs to be looked at and wasn't being called in this PR. Added the TODO into #31147 to track it.